### ISSUE TITLE
Original track tags vanish

### ISSUE THREAD

**Issue by WhitePeter** [kevinlekiller/mkvrg#4]
As I have noticed [here](https://github.com/mpv-player/mpv/issues/3405#issuecomment-239616609) already, mkvrg seems to make other tags, that were present before the treatment, disappear.

I do think, however, that this is an issue with mkvpropedit, because the man page seems to suggest that the original tags are kept. So this is more of an FYI. Or am I reading it wrong?

> With track mkvpropedit(1) will replace tags for a specific track. ___Additionally the tags read from filename will be assigned to the same track___. The track is specified in the same way edit selectors are specified (see below), e.g.  --tags track:a1:new-audio-tags.xml.

**Comment by kevinlekiller** [comment]
I think we can extract the tags, mkvtoolnix has some utility for that iirc, then we can possibly merge both XML's before using mkvpropedit.

**Comment by WhitePeter** [comment]
Yes, I think that would be possible, perhaps with the same mkvpropedit, but I really think this should be fixed in mkvpropedit. Or, at least I'd like to clear this up. Maybe it is just a misunderstanding, as I said.

**Comment by kevinlekiller** [comment]
Yeah, that's what I thought also from the man page. If you did not bring it up I would have not noticed!

bs1770gain seems to have a similar issue if you remux.

Here's a file before before running it through `bs1770gain -o . -a ~/Downloads/mkvtest/test5.mkv`:

http://sprunge.us/QgQI

Here it after:
http://sprunge.us/IAJE

All the subtitles are gone !

**Comment by WhitePeter** [comment]
Oh yes, forgot about that. I had the exact same issue the other day. I think that is because ffmpeg by default only selects one track of a kind. Subtitles seem to be a little different.
I think what bs1770gain should do at the remuxing stage is:
```
ffmpeg -i input -c copy -map 0 -map_metadata 0 -metadata RG_stuff output
```
Maybe they forgot the `-map 0`?

**Comment by WhitePeter** [comment]
Oh, and I now think I read it wrong with mkvpropedit, since filename must mean the xml file provided. Maybe mkvextract is the tool to get the original tags for the track.

**Comment by WhitePeter** [comment]
And it would be worth a try to simply use the option twice, once with the original tags.xml and then with those from mkvrg.

**Comment by kevinlekiller** [comment]
From what I can see, all the tags are extracted, not just specific tags per track.

**Comment by WhitePeter** [comment]
OK. FYI, I am currently checking if supplying two --tags options works.

**Comment by kevinlekiller** [comment]
It seems a bit complex;

mkvextract can output a XML with all the tags, it has a UID, but it doesn't have a track id which is compatible with bs1770gain or mkvpropedit.

mkvinfo has the uid, and a id which is compatible with bs1770gain also

**Comment by WhitePeter** [comment]
Hmm, I am still not ready for giving up on mkvpropedit. The man page does say replace or add. So there must be a way...

**Comment by kevinlekiller** [comment]
I'll run mkvpropedit through a debugger to see what's going on

**Comment by kevinlekiller** [comment]
Compiling mkvtoolnix takes a while, seems like a large project, not even sure if it accepted the CXXFLAGS even.

**Comment by WhitePeter** [comment]
maybe disable the GUI stuff. I had to anyway, since my QT is too old and I couldn't be bothered.
But it still did take some time.

**Comment by kevinlekiller** [comment]
Yeah, the GUI stuff is not compiling, I don't have the QT libs installed.

I think rake doesn't accept -j8 which is why it's so slow.

**Comment by WhitePeter** [comment]
Ah, and there is that. Could not figure that out myself, not that I really tried, though. I figured by the time I had found the equivalent for make -j4, rake would be done anyway. :)

**Comment by kevinlekiller** [comment]
Found it, `rake --multitask`, flying now

**Comment by WhitePeter** [comment]
Oh, thanks for that.

**Comment by kevinlekiller** [comment]
Looks CXXFLAGS were ignored: Reading symbols from src/mkvpropedit...(no debugging symbols found)...done.

**Comment by WhitePeter** [comment]
I am definitely out of my depth there. I have never even run a debugger, so can't help you there, sorry.

**Comment by kevinlekiller** [comment]
Alright, trying to compile mkvtoolnix with `CXXFLAGS="-O0 -ggdb3"` (these flags disable compile optimizations and enables most debugging symbols)  so gdb can step through the source code while it runs.

Basically you then run `gdb --args ~/mkvroolnix/src/mkvpropedit --tags track:a1:/tmp/tags.xml test.mkv`

And you can put a breakpoint on a function, for example `b display_update_element_result` https://github.com/mbunkus/mkvtoolnix/blob/master/src/propedit/propedit.cpp#L26

Then you type `step`

And you can run through the source code line by line as the program runs to see what is going on.

**Comment by kevinlekiller** [comment]
I should have read the manual haha :

https://github.com/mbunkus/mkvtoolnix/blob/06725a38daad6bcec2dbf331ef115409947fe67d/INSTALL#L75

**Comment by WhitePeter** [comment]
> Alright, trying to compile mkvtoolnix with CXXFLAGS="-O0 -ggdb3"

And it did not work? `./configure --help` says it is an environment variable that is used.

**Comment by WhitePeter** [comment]
> I should have read the manual haha

Looks like an odd way of setting those. I thought, either export them in the build environment or set them by some configure option. But then, what do I know? ;)

**Comment by WhitePeter** [comment]
LOL, second line of --help:

```
Usage: ./configure [OPTION]... [VAR=VALUE]...
```

**Comment by kevinlekiller** [comment]
> And it did not work? ./configure --help says it is an environment variable that is used.

It worked.

`Reading symbols from src/mkvpropedit...done.`

**Comment by WhitePeter** [comment]
Nice, happy debugging, I guess. :)

**Comment by kevinlekiller** [comment]
Want to know what's even weirder?

If you edit audio track 1 tags, then audio track 2 tags are erased also.

**Comment by WhitePeter** [comment]
Are you sure? I mean, how do they get back in then? After one run of mkvpropedit on only 1 track the other track tags are still there.

**Comment by kevinlekiller** [comment]
You're right, I tested again, not sure why the tags were not showing.

**Comment by kevinlekiller** [comment]
Going through the debugger I stumbled on this function: https://github.com/mbunkus/mkvtoolnix/blob/master/src/propedit/options.cpp#L221

Not sure if it works right however.

**Comment by WhitePeter** [comment]
I am sorry, but I really think I cannot be of much help there. I only have very rudimentary C skills, C++ is virtually non-existent in my head. :( Wouldn't even know where `m_targets` comes from and where it actually gets a value, for example.

As I said, I have very little coding skills in actual programming languages. I do know a thing or two about shell scripting, but that is mostly so I can avoid actual programming. ;)

**Comment by kevinlekiller** [comment]
That's fine, I'm posting an issue on the mkvtoolnix github issue tracker, as this is kind of complex for us to figure out.

**Comment by kevinlekiller** [comment]
The author replied, he wrote it is by design. So I suppose we need to find another way.

**Comment by WhitePeter** [comment]
Bummer. Well, on my side it is not very pressing, since I only use matroska for movies, mostly. I don't see the point in .mka, for example. So I really only need the RG tags on a per track basis. So what, if other tags disappear.
But I don't like such things, so eventually this should be fixed, but there is need to hurry. :)

**Comment by kevinlekiller** [comment]
Yeah, it bugs me too.

We can fix it, but we need to call mkvinfo on the file first to get the UID / track number of the track, then we need to run mkvextract to get the XML, in the XML, there's a UID, so we now know which track number it belongs to, then we edit that XML with the replaygain info and pass it to mkvpropedit, pretty complex.

**Comment by WhitePeter** [comment]
Do you think it makes sense to do this in the shell script, or maybe do this while fixing issue #3? There must be some facilities in python that make this less of a pain, right?

**Comment by kevinlekiller** [comment]
Yeah, we could probably make our own editor in python.

**Comment by WhitePeter** [comment]
Whoa, that sounds complicated. I was thinking more like, what you said with calling mkvinfo and mkvextract etc. but do that in python.
Before you told me about this script I was looking for XML parsing capabilities in python, just to get an idea. There are some. I did nothing more that see, if there are any. But this could make things easier with the actual XML editing I reckon.

**Comment by kevinlekiller** [comment]
There are already libraries for getting metadata from mkv for python (for example: https://pypi.python.org/pypi/enzyme/ )

Maybe we can edit one to pass the track id and get back a XML for the track we want.

**Comment by WhitePeter** [comment]
That'd be great. Why haven't I thought about looking for such libraries first?!
